### PR TITLE
Build torch.distributed with Gloo backend on macOS

### PR DESCRIPTION
**Pull request by pietern**
In facebookincubator/gloo#212, a libuv based Gloo transport was introduced,
which allows us to use Gloo on macOS (and later perhaps also Windows). This
commit updates CMake code to enable building with USE_DISTRIBUTED=1 on macOS.

A few notes:
* The Caffe2 ops are not compiled, for they depend on `gloo::transport::tcp`.
* The process group implementation uses `gloo::transport::tcp` on Linux (because of `epoll(2)` on Linux and `gloo::transport::uv` on macOS).
* The TCP store works but sometimes crashes on process termination.
* The distributed tests are not yet run.
* The nightly builds don't use `USE_DISTRIBUTED=1`.

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: For now, yes. I don't think there's much value in making this configurable. With libuv untested beyond the Gloo test suite, I rather not yet make it the default for Linux.

File: torch/lib/c10d/test/TestUtils.hpp
Comment (pietern) [reply torch/lib/c10d/test/TestUtils.hpp]: It is needed for the signal argument to `kill(2)` some place below.

**Review by mrshenli — APPROVED** [review]
LGTM! Approve to unblock. Please address test and lint failure before landing. They look relevant:

```
Aug 27 11:20:38 /Users/distiller/project/third_party/gloo/gloo/cuda_private.cu(102): error: incomplete type is not allowed
Aug 27 11:20:38 
Aug 27 11:20:38 1 error detected in the compilation of "/var/folders/ms/xg67k5sn16xc7sdr_w3q45840000gn/T//tmpxft_00005194_00000000-6_cuda_private.cpp1.ii".
Aug 27 11:20:38 CMake Error at gloo_cuda_generated_cuda_private.cu.o.Release.cmake:277 (message):
Aug 27 11:20:38   Error generating file
Aug 27 11:20:38   /Users/distiller/project/build/third_party/gloo/gloo/CMakeFiles/gloo_cuda.dir//./gloo_cuda_generated_cuda_private.cu.o
```

**Comment by ezyang** [comment]
This breaks OS X binary builds:

```
Sep 05 14:22:45 -- Checking for one of the modules 'libuv>=1.26'
Sep 05 14:22:45 CMake Error at /Users/distiller/project/conda/conda-bld/pytorch_1567693243716/_build_env/share/cmake-3.14/Modules/FindPkgConfig.cmake:701 (message):
Sep 05 14:22:45   None of the required 'libuv>=1.26' found
Sep 05 14:22:45 Call Stack (most recent call first):
Sep 05 14:22:45   third_party/gloo/cmake/Dependencies.cmake:46 (pkg_search_module)
Sep 05 14:22:45   third_party/gloo/CMakeLists.txt:56 (include)
Sep 05 14:22:45 
Sep 05 14:22:45 
Sep 05 14:22:45 FATALUnable to find static libuv library in 
Sep 05 14:22:45 CMake Error at third_party/gloo/cmake/Dependencies.cmake:56 (set_target_properties):
Sep 05 14:22:45   set_target_properties called with incorrect number of arguments.
Sep 05 14:22:45 Call Stack (most recent call first):
Sep 05 14:22:45   third_party/gloo/CMakeLists.txt:56 (include)
Sep 05 14:22:45 
Sep 05 14:22:45 
Sep 05 14:22:45 CMake Warning at cmake/Dependencies.cmake:1010 (message):
Sep 05 14:22:45   Metal is only used in ios builds.
Sep 05 14:22:45 Call Stack (most recent call first):
Sep 05 14:22:45   CMakeLists.txt:367 (include)
```

https://circleci.com/gh/pytorch/pytorch/2681548?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

**Comment by ezyang** [comment]
(I'm not reverting because these don't break PRs)

**Comment by ezyang** [comment]
I'm not sure but this may have also broken Android builds:

```
Sep 05 14:49:25 /opt/ndk/toolchains/llvm/prebuilt/linux-x86_64/lib/gcc/aarch64-linux-android/4.9.x/../../../../aarch64-linux-android/bin/ld: cannot find -lpthread
Sep 05 14:49:26 clang++: error: linker command failed with exit code 1 (use -v to see invocation)
Sep 05 14:49:26 caffe2/CMakeFiles/torch.dir/build.make:11236: recipe for target 'lib/libtorch.so' failed
Sep 05 14:49:26 make[2]: *** [lib/libtorch.so] Error 1
Sep 05 14:49:26 CMakeFiles/Makefile2:1393: recipe for target 'caffe2/CMakeFiles/torch.dir/all' failed
Sep 05 14:49:26 make[1]: *** [caffe2/CMakeFiles/torch.dir/all] Error 2
Sep 05 14:49:26 make: *** [all] Error 2
Sep 05 14:49:26 Makefile:138: recipe for target 'all' failed
```

https://circleci.com/gh/pytorch/pytorch/2681545?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

**Comment by pietern** [comment]
Thanks for flagging, @ezyang. Any idea why this didn't show up in the PRs CI?

The binary builds shouldn't be building with USE_DISTRIBUTED=1, as I only set that in `.jenkins/pytorch/macos-build.sh` and take care of the dependencies there. By default, USE_DISTRIBUTED should still be off for macOS.

**Comment by pietern** [comment]
I see -- the build showed up in the commit status but didn't run (`Rejecting wheel 3.6 cpu`).

**Pull request closed without merging**
